### DISCUSSION TITLE
Allow anonymous type objects that implement a "data" interface

### DISCUSSION THREAD

**Discussion by HazyFish** [dotnet/csharplang#7719] [Language Ideas] [answered]
## Motivation

Currently, the norm is to use classes / records to represent DTOs. However, using inheritance is very limited since multiple inheritance is not allowed. As a result, I have to frequently create interfaces for DTOs and some corresponding classes that simply implements all properties defined in the interface. This is not good because

- A lot of duplicate code
- If I drop a property in the interface, I have to remember to manually drop it in the corresponding classes as well.

## Proposal

The idea is to allow use interfaces to anonymous type object to implement interfaces so we can use interfaces to represent DTOs, like the following:

```cs
interface I
{
    string A { get; set; }
}

var i = new { A = "" } as I;
```

### Current Solution

```cs
interface IUser { /* props A, B, C */ }

abstract class User : IUser { /* props A, B, C */ }
class ExternalUser : User { /* props D, E */ }
class InternalUser : User { /* props F */ }
class AdminUser : User { /* props G */ }

interface IUserDbModel : IUser { /* props U, V, W, X, Y, Z */ }
class ExternalUserDbModel : ExternalUser, IUserDbModel  { /* props U, V, W, X, Y, Z */ }
class InternalUserDbModel : InternalUser, IUserDbModel  { /* props U, V, W, X, Y, Z */ }
class AdminUserDbModel : AdminUser, IUserDbModel  { /* props U, V, W, X, Y, Z */ }

var user = new AdminUserDbModel { /* ... */  };
```

### Desired Solution

```cs
interface User { /* props A, B, C */ }
interface ExternalUser : User { /* props D, E */ }
interface InternalUser : User { /* props F */ }
interface AdminUser : User { /* props G */ }

interface UserDbModel : IUser { /* props U, V, W, X, Y, Z */ }
interface ExternalUserDbModel : ExternalUser, UserDbModel { } 
interface InternalUserDbModel : InternalUser, UserDbModel { }
interface AdminUserDbModel : AdminUser, UserDbModel { }

var user = new { /* ... */  } as AdminUserDbModel;
```

In this way, we do not need to define props U, V, W, X, Y, Z in every DbModel DTOs.

**Comment by HaloFour** [comment, marked as answer]
See:

#4301
#4148
#2517
https://github.com/dotnet/roslyn/issues/13

**Reply by HazyFish** [reply]
@HaloFour Thank you so much for your prompt response!!! Do you happen to know any existing source generator based workarounds that can automatically add properties to classes that will implement the interfaces?
